pkg/ads/ads.go

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ads
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"strings"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+	mcp "istio.io/api/mcp/v1alpha1"
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pkg/adsc"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/model"
+	"kmesh.net/kmesh-coredns-plugin/pkg/dns"
+	"kmesh.net/kmesh-coredns-plugin/pkg/options"
+)
+
+type AdsController struct {
+	server    *dns.KmeshDNSServer
+	adsClient *adsc.ADSC
+	vip       string
+}
+
+func NewAdsController(cfg *options.BootstrapConfigs, dnsServer *dns.KmeshDNSServer) (*AdsController, error) {
+	c := &AdsController{
+		server: dnsServer,
+		vip:    cfg.VIP,
+	}
+	client, err := adsc.New(cfg.XDSAddress, &adsc.ADSConfig{
+		InitialDiscoveryRequests: configInitialRequests(),
+		ResponseHandler:          c,
+		Config: adsc.Config{
+			Meta: model.NodeMetadata{
+				Generator: "api",
+			}.ToStruct(),
+			GrpcOpts: []grpc.DialOption{
+				// Because we use the custom grpc options for adsc, here we should
+				// explicitly set transport credentials.
+				// TODO: maybe we should use the tls settings within ConfigSource
+				// to secure the connection between istiod and remote xds server.
+				grpc.WithTransportCredentials(insecure.NewCredentials()),
+			},
+		},
+	})
+	if err != nil {
+		log.Fatalf("failed to dial XDS %s %v", "controller.gateway-system.svc.cluster.local:15010", err)
+		return nil, err
+	}
+
+	c.adsClient = client
+	return c, err
+}
+
+func (c *AdsController) Start() error {
+	return c.adsClient.Run()
+}
+
+func (c *AdsController) Stop() {
+	c.adsClient.Close()
+}
+
+func configInitialRequests() []*discovery.DiscoveryRequest {
+	out := make([]*discovery.DiscoveryRequest, 0, 1)
+	out = append(out, &discovery.DiscoveryRequest{
+		TypeUrl: collections.ServiceEntry.GroupVersionKind().String(),
```

Comment:
Istiod doesnot respect this

Reply:
any suggestions for this, I did test it and works as expected

Reply:
maybe i misread it. Will take a look later.